qadence/ml_tools/callbacks/writer_registry.py

```diff
@@ -0,0 +1,375 @@
+        return self.writer
+
+    def close(self) -> None:
+        """Closes the TensorBoard writer."""
+        if self.writer:
+            self.writer.close()
+
+    def write(self, result: OptimizeResult) -> None:
+        """
+        Logs the results of the current iteration to TensorBoard.
+
+        Args:
+            result (OptimizeResult): The optimization results to log.
+        """
+        # Not writing loss as loss is available in the metrics
+        # if result.loss is not None:
+        #     self.writer.add_scalar("loss", float(result.loss), result.iteration)
+        if self.writer:
+            for key, value in result.metrics.items():
+                self.writer.add_scalar(key, value, result.iteration)
+
+    def log_hyperparams(self, hyperparams: dict) -> None:
+        """
+        Logs hyperparameters to TensorBoard.
+
+        Args:
+            hyperparams (dict): A dictionary of hyperparameters to log.
+        """
+        if self.writer:
+            self.writer.add_hparams(hyperparams, {})
+
+    def plot(
+        self,
+        model: Module,
+        iteration: int,
+        plotting_functions: tuple[PlottingFunction, ...],
+    ) -> None:
+        """
+        Logs plots of the model using provided plotting functions.
+
+        Args:
+            model (Module): The model to plot.
+            iteration (int): The current iteration number.
+            plotting_functions (tuple[PlottingFunction, ...]): Functions used
+                to generate plots.
+        """
+        if self.writer:
+            for pf in plotting_functions:
+                descr, fig = pf(model, iteration)
+                self.writer.add_figure(descr, fig, global_step=iteration)
+
+    def log_model(
+        self,
+        model: Module,
+        dataloader: Union[None, DataLoader, DictDataLoader],
+    ) -> None:
+        """
+        Logs the model.
+
+        Currently not supported by TensorBoard.
+
+        Args:
+            model (Module): The model to log.
+            dataloader (DataLoader | DictDataLoader | None): DataLoader to use
+                for model input.
+        """
+        logger.warning("Model logging is not supported by tensorboard. No model will be logged.")
+
+
+class MLFlowWriter(BaseWriter):
+    """
+    Writer for logging to MLflow.
+
+    Attributes:
+        run: The active MLflow run.
+        mlflow: The MLflow module.
+    """
+
+    def __init__(self) -> None:
+        self.run: Run
+        self.mlflow: ModuleType
+
+    def open(self, config: TrainConfig, iteration: int = None) -> ModuleType | None:
+        """
+        Opens the MLflow writer and initializes an MLflow run.
+
+        Args:
+            config: Configuration object containing settings for logging.
+            iteration (int, optional): The iteration step to start logging from.
+                Defaults to None.
+
+        Returns:
+            mlflow: The MLflow module instance.
+        """
+        import mlflow
```

Comment:
why the import inside?

Reply:
Moved outside. 

Kept it inside to avoid unnecessarily importing mlflow in case it was not needed.